addons/spreadsheet_account/models/account.py

```diff
@@ -110,22 +293,91 @@ def spreadsheet_fetch_debit_credit(self, args_list):
+
+    @api.readonly
+    @api.model
+    def spreadsheet_fetch_partner_balance(self, args_list):
+        """Fetch data for ODOO.PARTNER.BALANCE formulas
+        The input list looks like this:
+        [{
+            date_from: {
+                range_type: "year"
+                year: int
+            },
+            date_to: {
+                range_type: "year"
+                year: int
+            },
+            company_id: int
+            codes: str[]
+            include_unposted: bool
+            partner_ids: str[]
+        }]
+        """
+        return self._spreadsheet_fetch_data(args_list, ['balance'], default_accounts=True)
+
+    @api.readonly
+    @api.model
+    def _spreadsheet_fetch_data(self, args_list, fields, default_accounts=False):
+        if not args_list:
+            return []
+
+        self._pre_process_date_period_boundaries(args_list)
+        timeline = self._get_timeline(args_list)
+        all_accounts = self._get_all_accounts(args_list, default_accounts=default_accounts)
+        all_lines = self._get_all_lines(args_list, fields, timeline, all_accounts)
+
         results = []
         for args in args_list:
-            company_id = args["company_id"] or self.env.company.id
-            domain = self._build_spreadsheet_formula_domain(args)
-            MoveLines = self.env["account.move.line"].with_company(company_id)
-            [(debit, credit)] = MoveLines._read_group(domain, aggregates=['debit:sum', 'credit:sum'])
-            results.append({'debit': debit or 0, 'credit': credit or 0})
+            subcodes = tuple({subcode for subcode in args["codes"] if subcode})
+
+            if not subcodes and not default_accounts:
+                results.append({field: 0.0 for field in fields})
+                continue
+
+            company_id = args['company_id'] or self.env.company.id
+            states = ['posted', 'draft'] if args['include_unposted'] else ['posted']
+            periods = args['date_periods']
+            if subcodes:
+                accounts = all_accounts[company_id].filtered(lambda account: account.with_company(company_id).code.startswith(subcodes))
+            else:
+                accounts = all_accounts[company_id].filtered(lambda account: account.with_company(company_id).account_type in ['liability_payable', 'asset_receivable'])
+            partner_ids = args.get('partner_ids')
+
+            matching_keys = set()
+            for account in accounts:
+                # Initial balanced accounts are cumulated over the periods due to their nature. For that reason,
+                # we need to add all previous period values for that account as well.
+                past_periods = []
+                if account.with_company(company_id).include_initial_balance:
+                    past_periods = timeline[0:timeline.index(periods[0])]
+
+                matching_keys.update(set(filter(
```

Comment:
Isn't this a computational nightmare?
For each cell, you iterate on each account, and then on each key. So basically ~O(n^2) with n the number of cells.